test/dist_autograd_test.py

```diff
@@ -60,63 +78,141 @@ def test_autograd_context(self):
+        self.assertEqual(2, len(next_funcs))
+
+        # We should now hit t1 and t2 in the autograd graph.
+        self.assertEqual("torch::autograd::AccumulateGrad", next_funcs[0][0].name())
+        self.assertEqual(t1, next_funcs[0][0].variable)
+        self.assertEqual(0, next_funcs[0][1])
+        self.assertEqual("torch::autograd::AccumulateGrad", next_funcs[1][0].name())
+        self.assertEqual(t2, next_funcs[1][0].variable)
+        self.assertEqual(0, next_funcs[1][1])
+
+        # Test recv functions.
+        recv_functions = ctx._recv_functions()
+        self.assertEqual(1, len(recv_functions))
+        self.assertEqual(ret.grad_fn, list(recv_functions.values())[0])
+
+    # Host receives tensors and actually runs tensor operations, return tensor
+    # result
+    def _verify_send_recv_functions_in_tensor_run(self, ctx):
+        # Get the send function.
+        send_functions = ctx._send_functions()
+        self.assertEqual(1, len(send_functions))
+
+        # Verify next function is AddBackward0
+        next_funcs = list(send_functions.values())[0].next_functions
+        self.assertEqual(1, len(next_funcs))
+        add_backward_fn = next_funcs[0][0]
+        self.assertEqual("AddBackward0", add_backward_fn.name())
+
+        # Verify the next two functions are the same recv backward function.
+        next_funcs = add_backward_fn.next_functions
+        self.assertEqual(2, len(next_funcs))
+        self.assertEqual(
+            "torch::distributed::autograd::RecvRpcBackward", next_funcs[0][0].name()
+        )
+        self.assertEqual(
+            "torch::distributed::autograd::RecvRpcBackward", next_funcs[1][0].name()
+        )
+        self.assertEqual(next_funcs[0][0], next_funcs[1][0])
+
+    def _test_autograd_functions(self, fn):
         dst_rank = (self.rank + 1) % self.world_size
         with dist_autograd.context() as context_id:
             t1 = torch.ones(3, 3, requires_grad=True)
             t2 = torch.zeros(3, 3, requires_grad=True)
-            ret = rpc.rpc_sync("worker{}".format(dst_rank), torch.add, args=(t1, t2))
-            rpc.rpc_sync(
-                "worker{}".format(dst_rank), _set_rpc_done, args=(context_id,)
-            )
+            ret = rpc.rpc_sync("worker{}".format(dst_rank), fn, args=(t1, t2))
+            rpc.rpc_sync("worker{}".format(dst_rank),
+                         _set_rpc_done_from_prev_rank, args=(context_id,))
 
-            # Get send function.
+            self._verify_send_recv_functions_in_client(context_id, t1, t2, ret)
+
+            # We should have send/recv functions from the previous rank, get all
+            # contexts in this node to find them.
+
+            # Wait for the prev rank to be done with rpc.
+            while not prev_rank_rpc_done:
+                time.sleep(0.1)
+                pass
+
+            # Now verify the autograd graph.
+            ctx = dist_autograd._retrieve_context(prev_rank_context_id)
```

Comment:
This is not going to succeed with the changes in #27951, because it is possible that the `prev_rank` (i.e. the rank with context_id `prev_rank_context_id`) would have exited the autograd context manager in their thread, and thus destroyed the autograd context. Instead of using `time.sleep()` as above, I think we need to do something like `rpc.sync_rpc()`, though @pritamdamania87 mentioned that this is going away. Is there any other to wait for all outstanding RPCs to complete ?

Reply:
thanks for heads up, if your diff landed first, I will rebase on it. otherwise I will leave this implementation as it is right now, and your diff can rebase on mine and add the change properly. does it sound good to you?

Reply:
Sounds good!